feat: add option to remove prefixed dates from titles

donni106 opened the pull request:
- added option in importer config to be able to remove prefixed dates from title strings
  - used regex based on number counts, which is enough for our case, we do not want to validate dates, we just want to match their formats per "dd.mm.yyyy", where d and m could be a single number or the year just two
- added spec with mocked config and xml example
- removed poi_record stuff that do not belong to the code base which were created on project init

SVA-277

---

![Bildschirmfoto 2022-01-11 um 12 12 37](https://user-images.githubusercontent.com/1942953/148932543-5d4e6203-8f6d-4c38-8efa-3e17b9fe15d3.png)

donni106 committed:
feat: add option to remove prefixed dates from titles

- added option in importer config to be able to remove prefixed
  dates from title strings
  - used regex based on number counts, which is enough for our
    case, we do not want to validate dates, we just want to match
    their formats per "dd.mm.yyyy", where d and m could be a single
    number or the year just two
- added spec with mocked config and xml example
- removed poi_record stuff that do not belong to the code base
  which were created on project init

SVA-277

tonekk commented on app/models/record.rb:
```diff
@@ -122,6 +131,10 @@ def parse_author(xml_item)
     end
+
+    def remove_prefixed_date_in(text)
+      text.gsub(/^\d{1,2}.\d{1,2}.\d{2,4}:?\s*/, "")
```
You forgot to escape the `.` character.
Like this, it will also match `2910111994` for example.

This is the right regex here:
```ruby
text.gsub(/^\d{1,2}\.\d{1,2}\.\d{2,4}:?\s*/, "")
```

donni106 replied:
nice

donni106 replied:
done with fa1dea5

donni106 committed:
fix: escape `.` in regex for prefixed dates

SVA-277

tonekk commented on spec/models/record_spec.rb:
```diff
@@ -1,13 +1,82 @@
+      xml_item = xml_data.doc.xpath("//item")
+      allow_any_instance_of(Record).to receive(:feed) { feed_without_prefixed_date }
+      expect(record.send(:parse_content_title, xml_item)).to eq("FACHFORUM \"DASEINSVORSORGE\" IM KREISKULTURHAUS")
+    end
```
Personally, I would encapsulate the tests in `context`-blocks, and put better naming to the contexts and tests.
For example:

```ruby
describe "parse content title" do

  subject do
    allow_any_instance_of(Record).to receive(:feed) { feed }
    Record.new.send(:parse_content_title, xml_item) 
  end

  context "with prefixed date option enabled" do
    let(:feed) { "INSERT FEED WITH DATE OPTION HERE" }
  
    it "convertes xml to json" do
      expect(subject).to eq("STRING WITH DATE")
    end
  end
    
  context "without prefixed date option enabled" do
    let(:feed) { "INSERT FEED WITHOUT DATE OPTION HERE" }
  
    it "convertes xml to json and removes date" do
      expect(subject).to eq("STRING WITHOUT DATE")
    end
  end
```

PS: I don't really understand why this option needs to be disabled, you could just look if there is a date and remove it, if there is one.
Also what is this `feed` used for, is it the general parsing options set for the whole project?

donni106 replied:
>PS: I don't really understand why this option needs to be disabled, you could just look if there is a date and remove it, if there is one.
> Also what is this `feed` used for, is it the general parsing options set for the whole project?

it is set per importer source and can be activated for a particular one. we do not know if this is wanted in general.
the `feed` is one importer source. a project can have multiple `feed`s.
does this answer your question?

tonekk replied:
yes.

I looked into the code now and since `feed` is just and `attr_accessor`, maybe you might as well do this since it's more explicit and less confusing:

```ruby
subject do
  record = Record.new
  record.feed = feed
  record.send(:parse_content_title, xml_item)
end
```

tonekk replied:
We do not need `allow_any_instance_of` here, since we are only testing one instance.
This mocking seemed confusing to me.

donni106 replied:
thanks, optimized in 10f44b6

donni106 committed:
refactor: rspecs for `record` concerning date in titles

- created a `subject` that can be reused in several contexts
  that define different `feed`s for different results

SVA-277

donni106 commented on spec/models/record_spec.rb:
```diff
@@ -1,13 +1,88 @@
+      #     </item>
+      #   </channel>
+      # </rss>
+      xml_data = Nokogiri::XML::Builder.new do |xml|
+        xml.rss {
+          xml.channel {
+            xml.item {
+              xml.title "19.10.2021: FACHFORUM \"DASEINSVORSORGE\" IM KREISKULTURHAUS"
+              xml.link "https://www.seelow.de/news/index.php?news=683934"
+              xml.description "<img src=\"https://vorschau.verwaltungsportal.de/news/6/8/3/9/3/4/3497999300.jpg\" alt=\"FACHFORUM \"DASEINSVORSORGE\" IM KREISKULTURHAUS\" /><br /><br /> Am 11.10.2021 kamen ca. 50 Referendare aus der ganzen Bundesrepublik zusammen und verständigten sich zum Thema &quot;Daseinsvorsorge im ländlichen Raum&quot;. Nach Vorträgen u.a. von ... <br /><br />[<a href=\"htt ...
+              xml.pubDate "Tue, 19 Oct 2021 00:00:00 +0200"
+            }
+          }
+        }
+      end
```
would you move this out of the `subject`?

tonekk replied:
yes. I would use
```ruby
RSpec.describe Record, type: :model do
  let(:record) { Record.new }
  let(:xml_data) do
    xml_data = Nokogiri::XML::Builder.new do |xml|
    end
  end
```

then you can even omit the `record = Record.new` line in the subject.
As a rule of thumb: Keep `subject` as small as possible (in most cases 1 line should do), and put all `let` definitions in the outmost scope (if they are not tied to a specific context).

donni106 replied:
done in 9cf22e4

tonekk commented on spec/models/record_spec.rb:
```diff
@@ -1,13 +1,88 @@
+      #       <pubDate>Tue, 19 Oct 2021 00:00:00 +0200</pubDate>
+      #     </item>
+      #   </channel>
+      # </rss>
```
I feel like this comment is a bit verbose.
If you're in doubt that someone doesn't know what `Nokogiri` does, maybe put something like
```ruby
# Build XML dummy schema with nokogiri
```
above it, the rest should be clear imho.

donni106 replied:
done in 9cf22e4

tonekk replied:
Added the comment, but also left the large comment in place?

donni106 replied:
true, removed it with 33d30a1
